File: 课程资料和情报（oj上的AC代码）/简单的整数划分问题.cpp
//没看到是若干组数据。。。。
#include <bits/stdc++.h>
using namespace std; 
int n;
int dp[51][51];
int dfs(int x,int mmax){
    if(dp[x][mmax])return dp[x][mmax];
    if(x<=1||mmax==1)return dp[x][mmax]=1;
    int total=0;
    for(int i=min(mmax,x);i>=1;i--)total+=dfs(x-i,i);
    return dp[x][mmax]=total;
}
int main(){
    while(cin>>n)cout<<dfs(n,n)<<"\n";
    return 0;
}

File: 课程资料和情报（oj上的AC代码）/简单的整数划分问题.py
def dfs(i,m,dp):
    if i<=1:return 1
    m,ans=min(i,m),0
    if dp[i][m]:return dp[i][m]
    for j in range(m,0,-1):ans+=dfs(i-j,j,dp)
    dp[i][m]=ans
    return ans
while True:
    try:m=int(input())
    except:break
    dp=[[0 for j in range(m+5)] for i in range(m+5)]
    print(dfs(m,m,dp))
    